### PR TITLE
Unify GeometryOptimisationProblem and OptimisationProblem

### LINKED ISSUE #792: Make `GeometryOptimisationProblem` a sub-class of `OptimisationProblem`
This feels like an alternative API to set up a general optimisation problem. It does get around some of the nastiness of e.g. `GeometryOptimisationProblem`, because different sets of constraints can be specified as inputs rather than hard-coded. I don't have a problem at all with a different API, but I am a little concerned that we're starting to have lots of different ways of doing things.

_Originally posted by @CoronelBuendia in https://github.com/Fusion-Power-Plant-Framework/bluemira/pull/530#discussion_r769224251_

### PR DESCRIPTION
**Pull request by CoronelBuendia**
## Linked Issues

Does this PR close or fix any Issues? Remember to create an Issue before starting work so that your fix / proposal can be addressed by the team.

closes #792 

## Description

Makes `GeometryOptimisationProblem` a subclass of `OptimisationProblem`

## Interface Changes

API to `GeometryOptimisationProblem` `__init__` changes and `.solve` -> `.optimise`

`.parameterisation` -> `._parameterisation`

Makes some classical objective and constraint functions pre-packaged in `geometry/optimisation.py`

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Comment by CoronelBuendia** [comment]
@shaildesai96 some TF builder base class changes coming in here. Keep an eye out for when this gets merged in. I can help if anything breaks, but it should just be a couple of name changes.

File: bluemira/geometry/optimisation.py [hunk 1]
```diff
@@ -46,44 +121,54 @@ class GeometryOptimisationProblem(abc.ABC):
...
-    def __init__(self, parameterisation: GeometryParameterisation, optimiser: Optimiser):
-        self.parameterisation = parameterisation
-        self.optimiser = optimiser
-        self.optimiser.build_optimiser(parameterisation.variables.n_free_variables)
-        self.optimiser.set_lower_bounds(np.zeros(optimiser.n_variables))
-        self.optimiser.set_upper_bounds(np.ones(optimiser.n_variables))
-        self.optimiser.set_objective_function(self.f_objective)
+    def __init__(
+        self,
+        parameterisation: GeometryParameterisation,
```
Comment (alexanderianblair) [inline bluemira/geometry/optimisation.py:126]: Possibly worth calling this input `geometry_parameterisation` to make it a little clearer to a reader the difference in usage between this and _parameterisation (although I know that they will reference the same object after the super().__init__ call).

**Review by alexanderianblair — CHANGES_REQUESTED** [review]
Nothing major - nice to see the same framework being used.

Main points are to communicate the similarities between the implementation of `OptimisationProblem` here and in Equilibria

File: bluemira/geometry/optimisation.py [hunk 2]
```diff
@@ -23,18 +23,93 @@
...
+def calculate_length(vector, parameterisation):
+    """
+    Calculate the length of the parameterised shape for a given state vector.
+    """
+    parameterisation.variables.set_values_from_norm(vector)
+    return parameterisation.create_shape().length
 
 
-class GeometryOptimisationProblem(abc.ABC):
+def minimise_length(vector, grad, parameterisation, ad_args=None):
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/geometry/optimisation.py:53]: Yes, I did hesitate to put one in.. I have tried in the past to make a decorator that does this, but it got a bit fiddly (I can't remember why exactly). Perhaps we can do this in a separate PR.

[the remark this answers is not in the document]
Comment (alexanderianblair) [reply bluemira/geometry/optimisation.py:54]: Happy with that - this isn't a blocker, by any means

File: bluemira/geometry/optimisation.py [hunk 3]
```diff
@@ -23,18 +23,93 @@
...
+            calculate_signed_distance,
+            vector,
+            f0=constraint,
+            args=(parameterisation, n_shape_discr, koz_points),
+            **ad_args,
+        )
+    return constraint
+
+
+class GeometryOptimisationProblem(OptimisationProblem):
```
[the remark this answers is not in the document]
Comment (alexanderianblair) [reply bluemira/geometry/optimisation.py:113]: Sounds good - I think that's the clearest of both options, on balance

**Review by alexanderianblair — APPROVED** [review]
Happy now the comments have been addressed!

**Review by je-cook — APPROVED** [review]

**Review by hsaunders1904 — CHANGES_REQUESTED** [review]

File: bluemira/builders/EUDEMO/tf_coils.py [hunk 4]
```diff
@@ -237,13 +237,13 @@ def run(self):
...
         """
-        super().run(
+        parameterisation = super().run(
             params=self._params,
             wp_cross_section=self._wp_cross_section,
             separatrix=self._separatrix,
             keep_out_zone=self._keep_out_zone,
         )
-        self._centreline = self._design_problem.parameterisation.create_shape()
+        self._centreline = parameterisation.create_shape()
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/builders/EUDEMO/tf_coils.py:246]: Hmm, yeah I suppose the signature should be consistent. What are you suggesting here; just to return the parameterisation in this method? Or something else?

File: bluemira/geometry/optimisation.py [hunk 5]
```diff
@@ -23,18 +23,94 @@
...
 from bluemira.geometry.wire import BluemiraWire
-from bluemira.utilities.optimiser import Optimiser
+from bluemira.utilities.opt_problems import (
+    OptimisationConstraint,
+    OptimisationObjective,
+    OptimisationProblem,
+)
+from bluemira.utilities.optimiser import Optimiser, approx_derivative
+
+__all__ = ("GeometryOptimisationProblem", "minimise_length")
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/geometry/optimisation.py:43]: Yes, I've always seen it as a list too. But then I saw somebody using a tuple, and thought an immutable type makes more sense.

File: bluemira/builders/EUDEMO/tf_coils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hsaunders1904) [reply bluemira/builders/EUDEMO/tf_coils.py:246]: Would what `WallBuilder` does work here? 

I think the equivalent in this case would be:

```python
    super().run(
            params=self._params,
            wp_cross_section=self._wp_cross_section,
            separatrix=self._separatrix,
            keep_out_zone=self._keep_out_zone,
    )
    self._centreline = self._shape.create_shape()
```

[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/builders/EUDEMO/tf_coils.py:246]: @alexanderianblair felt that we should return the parameterisation from OptimisationProblem but we don't need to do that in the builder base class.

I can put `self._shape = self._design_problem.optimise()` in the base class

[the remark this answers is not in the document]
Comment (hsaunders1904) [reply bluemira/builders/EUDEMO/tf_coils.py]: Not sure I'm completely following. `OptimisationProblem.optimise()` is returning the paramterisation...

I'm happy with the state it's in now if you are. If you are happy, there's some commented out code in `bluemira.builders.shapes.OptimisedShapeBuilder.run` that should probably be removed

[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/builders/EUDEMO/tf_coils.py]: ah shit my bad, thanks

**Comment by hsaunders1904** [comment]
This has picked up some conflicts, but I'm happy once the CI is

**Pull request merged**